# Basic Absolute and Global Scaling Vignette for Kim et al 2022
# Here we go through absolute and global scaling for the scRNA-seq "Mixology" ground truth dataset by Tian et al
# This dataset contains both ERCCs and UMIs


# Import matrices
# Matrics were generated by processing raw data from GSE118767 through a standard scPipe pipeline
# More details on scPipe found here: https://github.com/LuyiTian/scPipe

umi_raw <- read.csv(file = paste0("UMI Gene Count.csv"), sep = ",", header = TRUE, row.names = 1)
ercc_raw <- read.csv(file = paste0("ERCC Gene Count.csv"), sep = ",", header = TRUE, row.names = 1)
barcodes <- read.csv(file = paste0("barcodes.csv"), sep = ",", header = TRUE, row.names = 1)

# UMI Global Scaling
# To globally scale UMI data, we perform a standard procedure of taking relative counts (RC) of all genes and subsequenting applying a log scaling step
# As ERCCs are not relevant for scaling with UMI data, we first remove them from the raw matrix
ERCC_index <- grep(pattern = "^ERCC-", x = rownames(umi_raw), value = FALSE)
umi_raw <- umi_raw[-ERCC_index, ]

# Create a Seurat object
UMI_mix <- CreateSeuratObject(counts = umi_raw)
UMI_mix <- AddMetaData(UMI_mix, metadata = barcodes$mRNA_amount, col.name = "mRNAConcentration") # some ground truth data
Idents(object = UMI_mix) <- UMI_mix @meta.data$mRNAConcentration
UMI_mix <- subset(UMI_mix, subset = nCount_RNA > 200) # filter out very low transcript cells

# RC + log scaling
UMI_mix <- NormalizeData(UMI_mix, normalization.method = "RC")
UMI_mix @assays$RNA @data <- as.matrix(log2(UMI_mix @assays$RNA @data + 1))
UMI_mix@meta.data$scaledcounts <- colSums(UMI_mix @assays$RNA @data)

scattercol <- c("seagreen4", "seagreen3", "seagreen2", "seagreen1")
FeatureScatter(UMI_mix, "scaledcounts", "mRNAConcentration", slot = "data", pt.size = 1.5, cols = scattercol) +
  theme(
    aspect.ratio = 1, text = element_text(size = 14), axis.text = element_text(size = 12),
    plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")
  ) +
  xlab("Transcripts") +
  ylab("mRNA (pg)") +
  stat_cor() +
  ggtitle(label = "UMI Global Scaling")

# UMI Absolute Scaling
# To perform absolute scaling, we use the raw reads by only applying the log scaling step
UMI_mix @assays$RNA @data <- as.matrix(log2(UMI_mix @assays$RNA @counts + 1)) # use the original count data for log scaling
UMI_mix @meta.data$scaledcounts <- colSums(UMI_mix @assays$RNA @data)

FeatureScatter(UMI_mix, "scaledcounts", "mRNAConcentration", slot = "data", pt.size = 1.5, cols = scattercol) +
  theme(
    aspect.ratio = 1, text = element_text(size = 14), axis.text = element_text(size = 12),
    plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")
  ) +
  xlab("Transcripts") +
  ylab("mRNA (pg)") +
  stat_cor() +
  ggtitle(label = "UMI Absolute Scaling")

# ERCC Absolute Scaling
# To perform scaling for ERCC datasets, we use tools built in the scran package for ERCC.
# Cells are scaled with computeSpikeFactors and subsequently log-scaled

# Create Seurat Object
ERCC_mix <- CreateSeuratObject(counts = ercc_raw)
ERCC_mix <- AddMetaData(ERCC_mix, metadata = barcodes$mRNA_amount, col.name = "mRNAConcentration")
ERCC_mix <- AddMetaData(ERCC_mix, PercentageFeatureSet(ERCC_mix, pattern = "^ERCC-"), col.name = "ERCC")
ERCC_mix <- subset(ERCC_mix, subset = nCount_RNA > 200 & ERCC > 0.1) # filter for minimum amount of ERCCs
RNA_conc <- ERCC_mix @meta.data$mRNAConcentration

filteredcounts <- GetAssayData(object = ERCC_mix, assay = "RNA", slot = "data") # pull transcript data from Seurat object and into SCE
scran.data <- SingleCellExperiment(assays = list(counts = filteredcounts))
is.spike <- grepl("^ERCC-", rownames(scran.data))
scran.data <- splitAltExps(scran.data, ifelse(is.spike, "ERCC", "gene")) # Defining ERCC, also removes them from the main data

# Compute a spike-in size factor for each cell computed from the sum of all spike-in transcripts
# Mean of all size factors will = 0
scran.data <- computeSpikeFactors(scran.data, "ERCC")
scran.data <- logNormCounts(scran.data, log = FALSE) # scale with spike-in size factors, log-scale later

# load scaled data back into seurat
ERCC_mix <- CreateSeuratObject(counts = as.sparse(x = assay(scran.data, "normcounts")))
ERCC_mix @meta.data$mRNAConcentration <- RNA_conc

ERCC_mix @assays$RNA @data <- as.matrix(log2(ERCC_mix @assays$RNA @counts + 1)) # apply log-scaling
Idents(object = ERCC_mix) <- ERCC_mix @meta.data$mRNAConcentration

FeatureScatter(ERCC_mix, "nCount_RNA", "mRNAConcentration", slot = "data", pt.size = 1.5, cols = scattercol) +
  theme(
    aspect.ratio = 1, text = element_text(size = 14), axis.text = element_text(size = 12),
    plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")
  ) +
  xlab("Transcripts") +
  ylab("mRNA (pg)") +
  stat_cor() +
  ggtitle(label = "ERCC Absolute Scaling")

# ERCC Global Scaling
# There are a number of ways to do this (within or out of scran), here we use the simple method of using relative counts within Seurat.
# For consistency, we will once again load data into SCE as before to remove ERCCs but not scale with size-factors
ERCC_mix <- CreateSeuratObject(counts = ercc_raw)
ERCC_mix <- AddMetaData(ERCC_mix, metadata = barcodes$mRNA_amount, col.name = "mRNAConcentration")
ERCC_mix <- AddMetaData(ERCC_mix, PercentageFeatureSet(ERCC_mix, pattern = "^ERCC-"), col.name = "ERCC")
ERCC_mix <- subset(ERCC_mix, subset = nCount_RNA > 200 & ERCC > 0.1)
RNA_conc <- ERCC_mix @meta.data$mRNAConcentration

filteredcounts <- GetAssayData(object = ERCC_mix, assay = "RNA", slot = "data")
scran.data <- SingleCellExperiment(assays = list(counts = filteredcounts))
is.spike <- grepl("^ERCC-", rownames(scran.data))
scran.data <- splitAltExps(scran.data, ifelse(is.spike, "ERCC", "gene")) # remove ERCCs, do not scale

# load scaled data back into seurat
ERCC_mix <- CreateSeuratObject(counts = as.sparse(x = assay(scran.data, "counts"))) # use normal counts
ERCC_mix @meta.data$mRNAConcentration <- RNA_conc

# Simple global scaling as previous described
ERCC_mix <- NormalizeData(ERCC_mix, normalization.method = "RC")
ERCC_mix @assays$RNA @data <- as.matrix(log2(ERCC_mix @assays$RNA @data + 1))
ERCC_mix@meta.data$scaledcounts <- colSums(ERCC_mix @assays$RNA @data)
Idents(object = ERCC_mix) <- ERCC_mix @meta.data$mRNAConcentration

FeatureScatter(ERCC_mix, "scaledcounts", "mRNAConcentration", slot = "data", pt.size = 1.5, cols = scattercol) +
  theme(
    aspect.ratio = 1, text = element_text(size = 14), axis.text = element_text(size = 12),
    plot.margin = margin(t = 0, r = 0, b = 0, l = 0, unit = "pt")
  ) +
  xlab("Transcripts") +
  ylab("mRNA (pg)") +
  stat_cor() +
  ggtitle(label = "ERCC Global Scaling")
